File: 1567-maximum-length-of-subarray-with-positive-product/1567-maximum-length-of-subarray-with-positive-product.java
class Solution {
	public int getMaxLen(int[] nums) {
		int L = nums.length,
		start = 0,
		  max = 0,
	   nCount = 0,
		 prev = -1;

		for(int end = 0;end < L;end++) {
			if(nums[end] == 0) {
				start = end+1;
				nCount = 0;
				prev = -1;
				continue;
			} else if(nums[end] < 0)
				++nCount;

			if(prev == -1 && nCount == 1 ) {
				prev = end + 1;
				continue;
			}

			if(nCount%2 == 0)
				max = Math.max(max, end - start + 1);
			else
				max = Math.max(max, end - prev + 1);
		}
		return max;
	}
}

File: 1567-maximum-length-of-subarray-with-positive-product/1567-maximum-length-of-subarray-with-positive-product.cpp
class Solution {
public:
    int getMaxLen(vector<int>& nums) {
        int n=nums.size();
        int res=0;
        int val=1;
        int start=0;
        bool firstNegative=false;
        int firstNegative_idx=0;
        for(int i=0;i<n;i++){
            if(nums[i]==0){
                val=1;
                start=i+1;
                firstNegative=false;
                continue;
            }
            
            if(nums[i]<0 and !firstNegative){
                firstNegative=true;
                firstNegative_idx=i;
            }
            
            if(nums[i]<0){
                val*=-1;
            }

            if(val>0){
                res=max(res,i-start+1);
            }else if(val<0 and firstNegative==true){
                res=max(res,i-firstNegative_idx);
            }
            
        }
        
        return res;
    }
};